File: backend/app/models/notifications.py
import uuid
from datetime import datetime, timezone

from sqlalchemy import JSON, DateTime, ForeignKey, Index, String
from sqlalchemy.orm import Mapped, mapped_column, relationship

from ..core.database import Base


NOTIFICATION_KINDS = (
    "twin_activity",
    "social",
    "release",
    "artist",
    "ama",
    "slate_save",
    "slate_message",
    "follow",
    "review_helpful",
    "hidden_gem",
)


class Notification(Base):
    __tablename__ = "notifications"

    id: Mapped[str] = mapped_column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    user_id: Mapped[str] = mapped_column(
        String, ForeignKey("users.id", ondelete="CASCADE"), index=True
    )
    kind: Mapped[str] = mapped_column(String)
    payload: Mapped[dict] = mapped_column(JSON)
    read_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc)
    )

    user = relationship("User", foreign_keys=[user_id])

    __table_args__ = (
        Index("ix_notifications_user_unread", "user_id", "read_at", "created_at"),
    )

File: frontend/src/types/notifications.ts
export type NotificationKind =
  | "twin_activity"
  | "social"
  | "release"
  | "artist"
  | "ama"
  | "slate_save"
  | "slate_message"
  | "follow"
  | "review_helpful"
  | "hidden_gem"
  | "orbit_request"
  | "orbit_accepted"
  | "cut_invite"
  | "film_recommend"
  | "dm_reaction";

export interface Notification {
  id: string;
  kind: NotificationKind;
  payload: Record<string, unknown>;
  readAt: string | null;
  createdAt: string;
}
